api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,309 @@
+}
+
+// SyncLoop will keep syncing with peers until catches up
+func (ss *EpochSync) SyncLoop(bc *core.BlockChain, worker *worker.Worker, isBeacon bool, consensus *consensus.Consensus) {
+	timeout := 2
+	for {
+		<-time.After(time.Duration(timeout) * time.Second)
+		timeout = ss.syncLoop(bc, worker, isBeacon, consensus)
+	}
+}
+
+func (ss *EpochSync) syncLoop(bc *core.BlockChain, worker *worker.Worker, isBeacon bool, _ *consensus.Consensus) (timeout int) {
+	maxHeight := getMaxPeerHeight(ss.syncConfig)
+	for {
+		block := bc.CurrentBlock()
+		height := block.NumberU64()
+		if height >= maxHeight {
+			utils.Logger().Info().
+				Msgf("[SYNC] Node is now IN SYNC! (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 60
+		}
+
+		utils.Logger().Info().
+			Msgf("[SYNC] Node is OUT OF SYNC (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+				isBeacon, bc.ShardID(), maxHeight, height)
+
+		var heights []uint64
+		curEpoch := block.Epoch()
+		for i := 0; i < int(SyncLoopBatchSize); i++ {
+			epochLastBlock := shard.Schedule.EpochLastBlock(curEpoch.Uint64())
+			if epochLastBlock > maxHeight {
+				break
+			}
+			if height != epochLastBlock {
+				heights = append(heights, epochLastBlock)
+			}
+			curEpoch = curEpoch.Add(curEpoch, common.Big1)
+		}
+
+		if len(heights) == 0 {
+			// We currently on top.
+			return 10
+		}
+
+		err := ss.ProcessStateSync(heights, bc, worker)
+		if err != nil {
+			utils.Logger().Error().Err(err).
+				Msgf("[SYNC] ProcessStateSync failed (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 2
+		}
+	}
+}
+
+// ProcessStateSync processes state sync from the blocks received but not yet processed so far
+func (ss *EpochSync) ProcessStateSync(heights []uint64, bc *core.BlockChain, worker *worker.Worker) error {
+
+	var payload [][]byte
+
+	ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
+		resp := peerConfig.GetClient().GetBlocksByHeights(heights)
+		if resp == nil {
+			return false
+		}
+		if len(resp.Payload) == 0 {
+			return false
+		}
+		payload = resp.Payload
+		return true
+	})
+
+	if len(payload) == 0 {
+		return errors.New("empty payload")
+	}
+
+	decoded := make([]*types.Block, 0, len(payload))
+	for idx, blockBytes := range payload {
+		block, err := RlpDecodeBlockOrBlockWithSig(blockBytes)
+		if err != nil {
+			return errors.Wrap(err, "failed decode")
+		}
+
+		if !block.IsLastBlockInEpoch() {
+			return errors.Errorf("block is not last index %d hash %s", idx, block.Hash().Hex())
+		}
+
+		decoded = append(decoded, block)
+	}
+
+	for _, block := range decoded {
+		sig, bitmap, err := chain.ParseCommitSigAndBitmap(block.GetCurrentCommitSig())
+		if err != nil {
+			return errors.Wrap(err, "parse commitSigAndBitmap")
+		}
+
+		// Signature validation.
+		err = bc.Engine().VerifyHeaderSignature(bc, block.Header(), sig, bitmap)
```

Comment:
Actually, this verification won't work, because the commit/sig stored in the block N is for the last block N-1, to get the commit/sig for block N, you need to find them in the block N+1.

Reply:
Have you tested this part of the code? I bet the verification of the sig should fail.